src/adaptors/amqp/qd_listener.c

```diff
@@ -113,20 +118,116 @@ static void handle_listener(pn_event_t *e, qd_server_t *qd_server, void *context
     }
 }
 
-qd_listener_t *qd_listener(qd_server_t *server)
+qd_listener_t *qd_listener_create(qd_dispatch_t *qd, qd_entity_t *entity)
 {
+    qd_error_clear();
+
     qd_listener_t *li = new_qd_listener_t();
-    if (!li) return 0;
+    if (!li) {
+        char *name = qd_entity_opt_string(entity, "name", "UNKNOWN");
+        qd_error(QD_ERROR_CONFIG, "Failed to create listener %s: resource allocation failed", name);
+        free(name);
+        return 0;
+    }
+
     ZERO(li);
+    DEQ_ITEM_INIT(li);
     sys_atomic_init(&li->ref_count, 1);
     sys_atomic_init(&li->connection_count, 0);
-    li->server      = server;
-    li->http = NULL;
+    li->server       = qd_dispatch_get_server(qd);
+    li->http         = NULL;
     li->type.context = li;
     li->type.handler = &handle_listener;
+
+    if (qd_server_config_load(&li->config, entity, true) != QD_ERROR_NONE) {
+        qd_log(LOG_CONN_MGR, QD_LOG_ERROR, "Unable to create listener: %s", qd_error_message());
+        qd_listener_decref(li);
+        return 0;
+    }
+
+    char *fol = qd_entity_opt_string(entity, "failoverUrls", 0);
+    if (fol) {
+        li->config.failover_list = qd_failover_list(fol);
+        free(fol);
+        if (li->config.failover_list == 0) {
+            qd_log(LOG_CONN_MGR, QD_LOG_ERROR, "Unable to create listener, bad failover list: %s",
+                   qd_error_message());
+            qd_listener_decref(li);
+            return 0;
+        }
+    } else {
+        li->config.failover_list = 0;
+    }
+
+    if (li->config.ssl_profile_name) {
+        li->tls_config = qd_tls_config(li->config.ssl_profile_name,
+                                       QD_TLS_TYPE_PROTON_AMQP,
+                                       QD_TLS_CONFIG_SERVER_MODE,
+                                       li->config.verify_host_name,
+                                       li->config.ssl_require_peer_authentication);
+        if (!li->tls_config) {
+            // qd_tls_config() sets qd_error_message():
+            qd_log(LOG_CONN_MGR, QD_LOG_ERROR, "Failed to configure TLS for Listener %s: %s",
+                   li->config.name, qd_error_message());
+            qd_listener_decref(li);
+            return 0;
+        }
+
+        li->tls_ordinal              = qd_tls_config_get_ordinal(li->tls_config);
+        li->tls_oldest_valid_ordinal = qd_tls_config_get_oldest_valid_ordinal(li->tls_config);
+        qd_tls_config_register_update_callback(li->tls_config, li,
```

Comment:
```suggestion
     if(li->tls_config)
        qd_tls_config_register_update_callback(li->tls_config, li,
```

Reply:
I was wondering if `qd_listener_create()` and `qd_listener_delete()` can be moved into its own commit so that this commit only contains the changes related to expiring listener side connection.